content/blog/2021/10/05/OptimizeYourPython.adoc

```diff
@@ -0,0 +1,438 @@
+
+[source,python]
+----
+
+
+@problem_fact
+class Timeslot:
+    def __init__(self, id, dayOfWeek, startTime, endTime):
+        self.id = id
+        self.dayOfWeek = dayOfWeek
+        self.startTime = startTime
+        self.endTime = endTime
+
+    @planning_id
+    def getId(self):
+        return self.id
+
+    def __str__(self):
+        return "Timeslot(id=" + str(self.id) + \
+               ", dayOfWeek=" + str(self.dayOfWeek) + ", startTime=" + str(self.startTime) + \
+               ", endTime=" + str(self.endTime) + ")"
+----
+
+=== Planning Entities
+
+Planning entities change throughout solving. Lesson is a planning entity, since its `room` and
+`timeslot` properties change throughout solving. Since the `room` and `timeslot` properties change throughout solving, they are known as planning variables. Let's create the `Lesson` class.
+
+[source,python]
+----
+from optapy import planning_entity, planning_variable
+
+@planning_entity
+class Lesson:
+    def __init__(self, id, subject, teacher, studentGroup, timeslot=None, room=None):
+        self.id = id
+        self.subject = subject
+        self.teacher = teacher
+        self.studentGroup = studentGroup
+        self.timeslot = timeslot
+        self.room = room
+
+    @planning_id
+    def getId(self):
+        return self.id
+
+    @planning_variable(Timeslot, ["timeslotRange"])
+    def getTimeslot(self):
+        return self.timeslot
+
+    def setTimeslot(self, newTimeslot):
+        self.timeslot = newTimeslot
+
+    @planning_variable(Room, ["roomRange"])
+    def getRoom(self):
+        return self.room
+
+    def setRoom(self, newRoom):
+        self.room = newRoom
+
+    def __str__(self):
+        return "Lesson(id=" + str(self.id) + \
+                ", timeslot=" + str(self.timeslot) + ", room=" + str(self.room) + \
+                ", teacher=" + str(self.teacher) + ", subject=" + str(self.subject) + \
+                ", studentGroup=" + str(self.studentGroup) + ")"
+----
+
+The `@planning_entity` decorator registers the class as a planning entity, which allows OptaPlanner to assign its planning variables and for it to be used in constraints.
+The `@planning_variable(variable_type, [value_range_provider_refs...])` decorator registers a method as the getter of a planning variable.
+The getter must be named `get<X>` and the setter must be named `set<X>`.
+The first argument, `variable_type`, tells OptaPlanner what type of values OptaPlanner can assign to this planning variable.
+The second argument, `value_range_provider_refs`, tells OptaPlanner what value ranges (more info on that later) it takes its values from.
+
+== Constraints
+
+Constraints tell OptaPlanner the score of the solution. A score represents the quality of a specific solution. The higher the better. OptaPlanner looks for the best solution, which is the solution with the highest score found in the available time. It might be the optimal solution.
```

Comment:
```suggestion
Constraints define the score calculation, or the fitness function, of a planning problem. Each solution of a planning problem is graded with a score. A score represents the quality of a specific solution. The higher the score the better. OptaPlanner looks for the best solution, which is the solution with the highest score found in the available time. It might or might not be the optimal solution.
```